iceoryx_utils/include/iceoryx_utils/internal/lockfree_queue/index_queue.hpp

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2019 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "iceoryx_utils/internal/lockfree_queue/buffer.hpp"
+#include "iceoryx_utils/internal/lockfree_queue/cyclic_index.hpp"
+#include "iceoryx_utils/internal/lockfree_queue/unique.hpp"
+
+#include <atomic>
+
+namespace iox
+{
+/// @brief lockfree queue capable of storing indices 0,1,... Capacity-1
+template <uint64_t Capacity, typename ValueType = uint64_t>
+class IndexQueue
+{
+  public:
+    class UniqueIndex : public unique<ValueType>

```

Comment:
Please inherit from ```cxx::optional<ValueType>``` instead of unique. The reasons are
- it does the same like unique
- it is very well tested, reviewed and already proven in use
- unique has undefined behavior and bugs

Reply:
It is also a bit more heavyweight for this case than I would like it to be. Will take a look, the unique construction may be removed altogether because it costs performance while not adding much (a bit "safety" in the internals) .